Feature/kas 3327 paginering op overzicht procedurestappen

sergiofenoll opened the pull request:
Adds pagination to subcases & reduces the number of backend calls made.

sergiofenoll committed:
Add pagination to subcases overview

sergiofenoll committed:
Remove subcases route

We no load in the subcases in the overview page (using pagination) and
again in the subcase/overview route to show sibling subcases

This last behaviour will change in the future (we will add pagination
there as well, but separate from the pagination of subcases/overview)

sergiofenoll committed:
Prevent multiple backend calls per subcase item

Previosuly, the subcase-item component was using `computed` properties
of the subcase model to side-load data (e.g. the phases belonging to a
subcase as well as the approval state of the subcase). This resulted in
multiple backend calls being made multiple times (presumably because
they got fired off again when the component re-rendered). To prevent
this, and keep the number of backend calls manageable, we side-load the
data using tasks which don't refire.

Where possible, the number of backend calls has also been reduced by
removing superfluous calls.

sergiofenoll committed:
Refactor phases computed property on subcase

Fetching `agenda-activity` here seems unnecessary, because the backend
microservice will check itself if such activities exist. So in a case
where there are no agenda-activities, we will still only make 1 request
to the backend, and when there are activities, we will make 1 request
instead of 2.

Also adds a comment to the subcase phases post processing function.

sergiofenoll commented on app/components/subcases/subcase-item.js:
```diff
@@ -89,6 +95,22 @@ export default class SubcaseItemSubcasesComponent extends Component {
 
+  @task
+  *loadSubcasePhases() {
+    this.phases = yield this.subcasesService.getSubcasePhases(this.args.subcase);
```
This loading task replaces a call to [`subcase.phases`](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/d9b7831ae7494bf5e052d8a62b02fe82c64cb699/app/models/subcase.js#L57).

sergiofenoll committed:
Check if meeting is final for subcase approval

sergiofenoll commented on app/components/subcases/subcase-item.js:
```diff
@@ -89,6 +95,27 @@ export default class SubcaseItemSubcasesComponent extends Component {
+  }
+
+  @task
+  *loadSubcaseIsApproved() {
```
This loading task replaces a call to [`subcase.approved`](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/d9b7831ae7494bf5e052d8a62b02fe82c64cb699/app/models/subcase.js#L79)

sergiofenoll commented on app/pods/cases/case/subcases/overview/controller.js:
```diff
@@ -5,6 +5,8 @@ import { action } from '@ember/object';
 export default class CasesCaseSubcasesOverviewController extends Controller {
   @tracked case;
+  @tracked page = 0;
+  @tracked size = 25;
```
The JIRA story's description specifies that we should show 20 subcases. I was a bit lazy and set it to 25 (because that's the closest value that's provided by `Auk::Pagination` by default). If showing 20 subcases instead of 25 is desired, this can be changed of course.

tdn replied:
Doesn't really matter, as long as they can change it. 25 is fine for me

sergiofenoll commented on app/services/subcases-service.js:
```diff
@@ -54,6 +54,10 @@ export default class SubcasesService extends Service {
             date: geplandeStart,
           });
           if (phaseData.decisionResultId) {
+            // Why don't we load in this data in the custom-subcases-service?
```
Maybe this shouldn't actually be a comment in the code, but just wrote it down somewhere when it occurred to me :sweat_smile:

tdn replied:
Good question... The performance already seems quite OK as it is though. So let's leave this in there for now.

ValenberghsSven replied:
Instead of a question in comments, you could create a Jira ticket and mention it here
`// KAS-xxxx load in this data in the custom-subcases-service instead`

Part of the audit last year also suggested to not have "TO DO" or question comments in code unless there is linked ticket available so we should try to avoid adding them.

tdn commented:
Works well and looks good to me. If Jenkins passes this'll be good to merge as far as I can see.

sergiofenoll committed:
Merge branch 'development' into feature/KAS-3327-paginering-op-overzicht-procedurestappen

sergiofenoll committed:
Merge branch 'development' into feature/KAS-3327-paginering-op-overzicht-procedurestappen

sergiofenoll committed:
Check against meeting being nullish

sergiofenoll committed:
Use router service to refresh parent route

Using the router service here because it's more explicit in what route
we're trying to refresh.

sergiofenoll committed:
Only show pagination if there is data to paginate

ValenberghsSven commented on package.json:
```diff
@@ -116,6 +116,7 @@
     "ember-qunit": "^5.1.5",
     "ember-resolver": "^8.0.3",
     "ember-responsive": "^3.0.6",
+    "ember-router-service-refresh-polyfill": "^1.0.1",
```
sneaky :D, was expecting this in a seperate PR.
Guess we'll just go for it now and plan refactor stories later.

sergiofenoll replied:
Had to choose between adding it in now, or reviving `cases/case/subcases/subcase/route.js` just to have some kind of refresh action (and then also make sure it doesn't have name collissions, make sure it's being called correctly from the route, ...) so I thought just adding it was the easiest solution :P

ValenberghsSven commented on app/pods/cases/case/subcases/subcase/overview/route.js:
```diff
@@ -9,7 +9,16 @@ export default class CasesCaseSubcasesSubcaseOverviewRoute extends Route {
   async afterModel(model) {
     // For showing the history of subcases within this route, we need a list of subcases without the current model
-    const allSubcases = this.modelFor('cases.case.subcases');
+    //  We want to sort descending on date the subcase was concluded.
+    //  In practice, reverse sorting on created will be close
+    const allSubcases = await this.store.query('subcase', {
+      filter: {
+        case: {
+          id: this.case.id,
+        },
+      },
+      sort: '-created',
+    });
```
I get that the pagination was not compatible with the current setup where a parent model from route `cases/case/subcase` is passed down to this afterModel.
Querying these here will impact the performance of the subcase detail overview?
use-case:
Case with 10 subcases, I want to view all of the info
before PR: loaded all subcases once when loading the parent route, opening detail view 10 times by going back to the overview and opening another subcase just passes on that model.
after PR: We query subcases on the overview route, we open the first subcase and we need to query all subcases again. We go back to the overview and have to query all subcases, opening second subcase is another query to all subcases. and so on.

Am I seeing this wrong?
If I'm not, perhaps it would be a good idea to keep the call in the parent route to avoid doing it multiple times.
The queries will be cached after the first time, but cache invalidations happen often on active cases/subcases.

sergiofenoll replied:
The problem with keeping it in the parent route is that we force the `subcase` rout to only show whatever page we're on in the `subcases` route. Additionally, in a follow-up story we add pagination to the `subcase` route too, which should be separate from the pagination in the `subcases` route.

I don't know what the real life use case is like though, if there will be tons of `cases` with >25 `subcases`, so I don't know how often pagination is actually going to be used here.

ValenberghsSven replied:
I'll keep the possible performance hit in mind going forward.
Now that I know this is mainly a legacy data issue I guess querying the same thing a few times will be fine in terms of performance on average cases.

ValenberghsSven commented on app/pods/cases/case/subcases/subcase/controller.js:
```diff
@@ -1,9 +1,12 @@
   @action
   refreshSubcases() {
-    this.send('refreshParentModel');
+    this.router.refresh('cases.case.subcases');
```
this only made sense when the subcases were in the model of `cases.case.subcases`, if we keep the current approach, you would just need to refresh the current route `cases.case.subcases.subcase`

ValenberghsSven commented on app/components/subcases/subcase-item.js:
```diff
@@ -89,6 +95,27 @@ export default class SubcaseItemSubcasesComponent extends Component {
+
+  @task
+  *loadSubcaseIsApproved() {
+    const meeting = yield this.args.subcase.requestedForMeeting;
+    if (meeting?.isFinal) {
+      this.approved = !!(yield this.store.queryOne('agenda-item-treatment', {
+        'filter[subcase][id]': this.args.subcase.id,
+        'filter[decision-result-code][:id:]': [
+          CONSTANTS.DECISION_RESULT_CODE_IDS.GOEDGEKEURD,
+          CONSTANTS.DECISION_RESULT_CODE_IDS.KENNISNAME,
+        ].join(','),
+      }));
+    } else {
+      this.approved = false;
+    }
+  }
```
This aims to replace the computed property `approved` on subcase.js, which i'm fine with.
We use it in more places, so eventually this could maybe move to a util instead, good for now atleast.

The computed `phases` on subcase was listening to the `approved` computed, which is now being called less so I found an instance where the computed `phases` was not updated (refresh of page fixed it)
Have a subcase on a design agenda (only 1 or 2 phases, "ingediend op" en "geagendeerd op")
in the same browser window:
- look at the subcase detail
- go to agenda, set formal ok and approve and close the agenda
- go back to subcase detail, the phases should now have "goedgekeurd op", but they don't, only after refresh.

Possible solution:
add `this.phases = yield this.subcasesService.getSubcasePhases(this.args.subcase);` to the afterModel hook of `cases/case/subcases/subcase/` and pass on that args to `SubcaseDescriptionView`, or load it that component I guess.
Then the `phases` computed should no longer be referenced.

sergiofenoll replied:
The changes in this PR shouldn't change the behaviour you described :thinking: 
The detail view of a subcase `cases.case.subcases.subcase.overview` is still using the computed `phases` & `approved` properties of the `subcase` model. So, unless I'm mistaken, in your example you're going form subcase overview to agenda back to subcase overview, so whatever is happening has nothing to do with the changes here (because those happen in `cases.case.subcases.overview`, and you're never passing that page.

ValenberghsSven replied:
hmm, had to verify on dev branch and seems like you are right.
The problem is already present, my bad.

Could do the fix in a bug ticket since I think that fix will work, but ignore here

ValenberghsSven commented on app/pods/cases/case/subcases/overview/controller.js:
```diff
@@ -13,6 +15,24 @@ export default class CasesCaseSubcasesOverviewController extends Controller {
 
   @action
   refreshSubcases() {
-    this.send('refreshParentModel');
+    this.send('refreshSubcasesRoute');
+  }
```
why send when we can router with the polyfill? :D

sergiofenoll replied:
hah, good catch!

ValenberghsSven commented:
Managed to fix tests for now but still waiting on CSS, could we copy some of the needed classes into auk_additions like we did in other PR's?

sergiofenoll commented:
> Managed to fix tests for now but still waiting on CSS, could we copy some of the needed classes into auk_additions like we did in other PR's?

Is this comment for a different PR? No CSS changes were made in this PR :thinking:

ValenberghsSven commented:
> > Managed to fix tests for now but still waiting on CSS, could we copy some of the needed classes into auk_additions like we did in other PR's?
> 
> Is this comment for a different PR? No CSS changes were made in this PR thinking

yep it is....  ignore this.

sergiofenoll committed:
Add KAS ticket number to comment

sergiofenoll committed:
Use router service for refreshing in subcases too

ValenberghsSven commented:
I was going to comment something, started querying for some data to answer the question "I don't know what the real life use case is like though, if there will be tons of cases with >25 subcases, so I don't know how often pagination is actually going to be used here."

But it seems like I found anomalies in the data.
I tried querying to get a count of cases who had more than 20 subcases.
The data surprised me, I found number as high as 381!.
After digging deeper, something seems wrong with legacy data.
the number of 381 subcases is the case from the Jira ticket (which should only have 122?). I queried all cases, then their subcases (with the relation `dossier:doorloopt`)
Only 122 have a type `dossier:Procedurestap`, so the other 259 are some legacy conversion error.

Talked with Johan, subcases were incorrectly linked to 1 case in the original conversion.

It's out of scope but just wanted to indicate what I just spend some time on to find out ...

ValenberghsSven commented:
FYI, adjusted my query and from what I can tell:
Without looking at legacy data, the current highest amount of subcases in a case is 8 since the start of kaleidos.

ValenberghsSven reviewed (approved):
fine as it is now if jenkins is green.

edit: 1 test is failing on dev "Vertrouwelijk" in a pill not found, ignore here

ValenberghsSven commented on app/pods/cases/case/subcases/overview/template.hbs:
```diff
@@ -6,11 +6,41 @@
+      </Toolbar.Group>
+      <Toolbar.Group @position="right" as |Group|>
+        <Group.Item>
+          {{t "amount-showed"}}
```
forgot about this, "dossiers per pagina" => "procedurestappen per pagina"?
Might be confusing

sergiofenoll replied:
thanks! made it "Aantal resultaten per pagina", same thing we use in agendas overview

ValenberghsSven commented on app/pods/cases/case/subcases/overview/template.hbs:
```diff
@@ -6,11 +6,41 @@
+      </Toolbar.Group>
+      <Toolbar.Group @position="right" as |Group|>
+        <Group.Item>
+          {{t "amount-showed"}}
```
forgot about this, "dossiers per pagina" => "procedurestappen per pagina"?

sergiofenoll committed:
Use correct string in pagination of subcases

erikap commented on app/config/constants.js:
```diff
@@ -55,6 +55,10 @@ export default {
     KENNISNAME: 'http://kanselarij.vo.data.gift/id/concept/beslissings-resultaat-codes/9f342a88-9485-4a83-87d9-245ed4b504bf',
     INGETROKKEN: 'http://kanselarij.vo.data.gift/id/concept/beslissings-resultaat-codes/453a36e8-6fbd-45d3-b800-ec96e59f273b',
   },
+  DECISION_RESULT_CODE_IDS: {
+    GOEDGEKEURD: '56312c4b-9d2a-4735-b0b1-2ff14bb524fd',
+    KENNISNAME: '9f342a88-9485-4a83-87d9-245ed4b504bf',
+  },
```
Sorry, late to the review party, but I think this is a dangerous precedent to end up again in an unstructured config file with a mix of ids and URIs. We only consider URIs to be stable. 

I get that you need IDs to filter on multiple values in mu-cl-resources, but I would configure URIs here and get their IDs in the component where you need them. By using [`store.findRecordByUri`](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/development/app/services/store.js#L29) we ensure only 1 backend call is made per concept so performance impact is minimal.

ValenberghsSven replied:
hmm, kinda passed over this but we should try to avoid going back to ID's indeed.
We can filter on uri in the query in `subcase-item.js` (we did that in the computed in subcase.js)
```
 'filter[decision-result-code][:uri:]': [
          CONSTANTS.DECISION_RESULT_CODE_URIS.GOEDGEKEURD,
          CONSTANTS.DECISION_RESULT_CODE_URIS.KENNISNAME,
        ].join(','),
```

ValenberghsSven replied:
@sergiofenoll already made the change to start jenkins. Perhaps we can get this PR done for ACC release

sergiofenoll replied:
mu-cl-resources support zo'n "or" niet voor URIs, maar ik push zo een change waar we de IDs gewoon binnenhalen

ValenberghsSven replied:
"mu-cl-resources support zo'n "or" niet voor URIs" echt? Ik twijfelde even, had het beter eerst lokaal getest...

sergiofenoll replied:
ja... ik weet dat ook alleen omdat ik het gisteren in rocket chat had gevraagd (heb ook eerst zo'n query proberen te schrijven, maar dat gaat dus niet)
https://chat.semte.ch/channel/mu-semtech?msg=2SQi8azvmuDs3NLie
maar hopelijk wordt support ervoor at some point toegevoegd :triumph:

ValenberghsSven replied:
dat heb je met snel willen zijn op het einde van de dag :sweat_smile:  sorry..

ValenberghsSven committed:
Merge branch 'development' of github.com:kanselarij-vlaanderen/kaleidos-frontend into feature/KAS-3327-paginering-op-overzicht-procedurestappen

ValenberghsSven committed:
KAS-3317 use uri instead of id's

sergiofenoll committed:
Get ids from records instead of constants file
